### PR TITLE
[gemini] gemini support extra-dp

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#5042 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

gemini support extra-dp

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/plugin/gemini_plugin.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 from torch.optim import Optimizer
 from torch.optim.lr_scheduler import _LRScheduler as LRScheduler
 from torch.utils.data import DataLoader
+from torch.nn.parallel import DistributedDataParallel as DDP
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:13]: Don't use DDP. There sychronization logic may have conflicts

File: colossalai/zero/gemini/chunk/chunk.py [hunk 2]
```diff
@@ -61,12 +61,14 @@ class Chunk:
...
         chunk_size: int,
-        process_group: ProcessGroup,
+        zero_group: ProcessGroup,
         dtype: torch.dtype,
         init_device: Optional[torch.device] = None,
         cpu_shard_init: bool = False,
         keep_gathered: bool = False,
         pin_memory: bool = False,
+        ddp_group: ProcessGroup = None,
+        use_ddp: bool = False,
```
Comment (ver217) [inline colossalai/zero/gemini/chunk/chunk.py:71]: Only one new argument is needed. If `ddp_group` is None, ddp won't be used.

File: colossalai/booster/plugin/gemini_plugin.py [hunk 3]
```diff
@@ -458,7 +483,11 @@ def configure(
                 shardformer = ShardFormer(self.shard_config)
                 model, _ = shardformer.optimize(model)
 
-            model = GeminiDDP(model, **self.gemini_config, process_group=self.dp_group, verbose=self.verbose)
+            if self.use_ddp:
+                model = model.cuda()
+                model = DDP(model, process_group=self.ddp_group, **self.ddp_config)
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:488]: No need to wrap DDP.

**Review by ver217 — CHANGES_REQUESTED** [review]

File: colossalai/zero/gemini/chunk/chunk.py (hunk 2 above)
Comment (flybird11111) [reply colossalai/zero/gemini/chunk/chunk.py:71]: ddp_group is obtained from the group mesh and it never return None. use_ddp is required.

File: colossalai/booster/plugin/gemini_plugin.py (hunk 1 above)
Comment (flybird11111) [reply colossalai/booster/plugin/gemini_plugin.py]: Updated. DDP is removed. I ran forward within the 'no_sync' context, It seems that DDP is redundant.

File: colossalai/booster/plugin/gemini_plugin.py (hunk 3 above)
Comment (flybird11111) [reply colossalai/booster/plugin/gemini_plugin.py]: updated.

File: colossalai/booster/plugin/gemini_plugin.py [hunk 4]
```diff
@@ -349,6 +350,7 @@ def __init__(
         norm_type: float = 2.0,
         enable_tensor_parallelism: bool = False,
         tp_size: int = 1,
+        ddp_size:int = 1,
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:353]: Extra-dp-size may be clearer. By the way you shoud update docstring

File: colossalai/booster/plugin/gemini_plugin.py [hunk 5]
```diff
@@ -403,11 +405,13 @@ def __init__(
...
-        assert self.dp_size > 1, f"The size of the DP group should be greater than 1. Please reduce the TP group size."
-        self.pg_mesh = ProcessGroupMesh(self.dp_size, self.tp_size)
-        self.dp_group = self.pg_mesh.get_group_along_axis(DP_AXIS)
+        self.ddp_size = ddp_size
+        self.use_ddp = self.ddp_size > 1
+        self.zero_size = dist.get_world_size() // (self.tp_size * self.ddp_size)
+        self.pg_mesh = ProcessGroupMesh(self.zero_size, self.tp_size, self.ddp_size)
+        self.zero_group = self.pg_mesh.get_group_along_axis(ZERO_AXIS)
         self.tp_group = self.pg_mesh.get_group_along_axis(TP_AXIS)
+        self.ddp_group = self.pg_mesh.get_group_along_axis(DDP_AXIS)
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:414]: You'd better not create new process group if extra-dp size is `1`, as creating process group costs a lot of time. Same for tp

File: colossalai/booster/plugin/gemini_plugin.py [hunk 6]
```diff
@@ -403,11 +405,13 @@ def __init__(
         self.verbose = verbose
 
         self.tp_size = tp_size if self.enable_tensor_parallelism else 1
-        self.dp_size = dist.get_world_size() // self.tp_size
-        assert self.dp_size > 1, f"The size of the DP group should be greater than 1. Please reduce the TP group size."
-        self.pg_mesh = ProcessGroupMesh(self.dp_size, self.tp_size)
-        self.dp_group = self.pg_mesh.get_group_along_axis(DP_AXIS)
+        self.ddp_size = ddp_size
+        self.use_ddp = self.ddp_size > 1
+        self.zero_size = dist.get_world_size() // (self.tp_size * self.ddp_size)
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:410]: You should check if it's divisible

File: tests/test_checkpoint_io/test_gemini_checkpoint_io.py [hunk 7]
```diff
@@ -154,7 +155,7 @@ def run_dist(rank, world_size, port):
 
 
 @pytest.mark.dist
-@pytest.mark.parametrize("world_size", [4])
+@pytest.mark.parametrize("world_size", [8])
```
Comment (ver217) [inline tests/test_checkpoint_io/test_gemini_checkpoint_io.py:158]: Only test marked as largedist can use 8 processes

**Review by ver217 — CHANGES_REQUESTED** [review]

File: colossalai/booster/plugin/gemini_plugin.py [hunk 8]
```diff
@@ -402,12 +399,32 @@ def __init__(
...
-        self.pg_mesh = ProcessGroupMesh(self.dp_size, self.tp_size)
-        self.dp_group = self.pg_mesh.get_group_along_axis(DP_AXIS)
-        self.tp_group = self.pg_mesh.get_group_along_axis(TP_AXIS)
+        self.tp_size = tp_size
+        self.extra_dp_size = extra_dp_size
+        world_size = dist.get_world_size()
+        self.zero_size = world_size // (self.tp_size * self.extra_dp_size)
+        assert world_size == (self.tp_size * self.extra_dp_size) * self.zero_size, f"The global group size can't be evenly divided by the subgroup size."
+        assert self.zero_size > 1, f"Gemini's group size should be greater than 1, please reduce tensor paralllelism group size or extra-dp group size."
+        if self.tp_size == 1 and self.extra_dp_size == 1:
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:408]: This is complicated. Actually, creating an instance of ProcessGroupMesh won't take too long. 
Here is pseudo-code:
```python
DP_AXIS, ZERO_AXIS, TP_AXIS = 0, 1, 2
self.pg_mesh = ProcessGroupMesh(self.extra_dp_size, self.zero_size, self.tp_size)
self.zero_group = self.pg_mesh.get_group_along_axis(ZERO_AXIS)
self.extra_dp_group = self.pg_mesh.get_group_along_axis(DP_AXIS) if self.extra_dp_size > 1 else None
self.tp_group = self.pg_mesh.get_group_along_axis(TP_AXIS) if self.tp_size > 1 else None
```

Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:408]: This is complicated. Actually, creating an instance of ProcessGroupMesh won't take too long.
Here is pseudo-code:
```python
DP_AXIS, ZERO_AXIS, TP_AXIS = 0, 1, 2
self.pg_mesh = ProcessGroupMesh(self.extra_dp_size, self.zero_size, self.tp_size)
self.zero_group = self.pg_mesh.get_group_along_axis(ZERO_AXIS) if self.zero_size > 1 else _get_default_group()
self.extra_dp_group = self.pg_mesh.get_group_along_axis(DP_AXIS) if self.extra_dp_size > 1 else None
self.tp_group = self.pg_mesh.get_group_along_axis(TP_AXIS) if self.tp_size > 1 else None
```

File: colossalai/booster/plugin/gemini_plugin.py [hunk 9]
```diff
@@ -402,12 +399,32 @@ def __init__(
...
-        assert self.dp_size > 1, f"The size of the DP group should be greater than 1. Please reduce the TP group size."
-        self.pg_mesh = ProcessGroupMesh(self.dp_size, self.tp_size)
-        self.dp_group = self.pg_mesh.get_group_along_axis(DP_AXIS)
-        self.tp_group = self.pg_mesh.get_group_along_axis(TP_AXIS)
+        self.tp_size = tp_size
+        self.extra_dp_size = extra_dp_size
+        world_size = dist.get_world_size()
+        self.zero_size = world_size // (self.tp_size * self.extra_dp_size)
+        assert world_size == (self.tp_size * self.extra_dp_size) * self.zero_size, f"The global group size can't be evenly divided by the subgroup size."
+        assert self.zero_size > 1, f"Gemini's group size should be greater than 1, please reduce tensor paralllelism group size or extra-dp group size."
```
Comment (ver217) [inline colossalai/booster/plugin/gemini_plugin.py:407]: `zero_size==1` is allowed

File: colossalai/zero/gemini/chunk/chunk.py [hunk 10]
```diff
@@ -90,9 +91,11 @@ def __init__(
         self.chunk_size = chunk_size
         self.utilized_size = 0
 
-        self.torch_pg = process_group
+        self.torch_pg = zero_group
         self.pg_size = dist.get_world_size(self.torch_pg)
         self.pg_rank = dist.get_rank(self.torch_pg)
+        self.extra_dp_group = extra_dp_group
+        self.extra_dp_size = dist.get_world_size(self.extra_dp_group)
```
Comment (ver217) [inline colossalai/zero/gemini/chunk/chunk.py:98]: If `extra_dp_group` is None, then `extra_dp_size` is global world size, which is wrong

File: tests/test_shardformer/test_hybrid_parallel_grad_clip_norm/test_naive_optimizer.py [hunk 11]
```diff
@@ -102,28 +102,11 @@ def check_forward_backward(model_fn, data_gen_fn, output_transform_fn, loss_fn,
...
-            "use_lazy_init": False,
-            "precision": "fp32",
-            "max_norm": 5,
-        },
         {
             "tp_size": 2,
             "pp_size": 2,
             "num_microbatches": 4,
-            "enable_all_optimization": True,
+            "enable_all_optimization": False,
```
Comment (ver217) [inline tests/test_shardformer/test_hybrid_parallel_grad_clip_norm/test_naive_optimizer.py:109]: Why?

**Review by ver217 — CHANGES_REQUESTED** [review]

File: tests/test_shardformer/test_hybrid_parallel_grad_clip_norm/test_naive_optimizer.py (hunk 11 above)
Comment (flybird11111) [reply tests/test_shardformer/test_hybrid_parallel_grad_clip_norm/test_naive_optimizer.py:109]: This parameter is unrelated to the test 'grad_clip_norm', so I have set it to false.

File: colossalai/booster/plugin/gemini_plugin.py (hunk 4 above)
Comment (flybird11111) [reply colossalai/booster/plugin/gemini_plugin.py]: updated

File: colossalai/booster/plugin/gemini_plugin.py (hunk 8 above)
Comment (flybird11111) [reply colossalai/booster/plugin/gemini_plugin.py]: updated

File: colossalai/zero/gemini/chunk/chunk.py (hunk 10 above)
Comment (flybird11111) [reply colossalai/zero/gemini/chunk/chunk.py]: updated.

File: colossalai/booster/plugin/gemini_plugin.py (hunk 8 above)
Comment (ver217) [reply colossalai/booster/plugin/gemini_plugin.py]: It's not exactly fixed

**Review by ver217 — APPROVED** [review]

**Pull request merged**